File: task2/track1_run/track1/students/solution.cpp
#include <algorithm>
#include <cmath>
#include <vector>

#include <omp.h>

struct Rating {
    int user;
    int item;
    float rating;
};

class IncrementalSVD {
public:
    void load_base_model(float* user_matrix, float* item_matrix, int u_size, int i_size, int dim, float mean) {
        users = u_size;
        items = i_size;
        latent_dim = dim;
        global_mean = mean;
        P = user_matrix;
        Q = item_matrix;
        seen.clear();
    }

    void update(const std::vector<Rating>& incremental_batch) {
        seen.insert(seen.end(), incremental_batch.begin(), incremental_batch.end());
        if (seen.empty()) {
            return;
        }

        // Sparse full-batch reference: seen is the non-zero COO rating list.
        // It performs deterministic SGD over all observed incremental ratings.
        const float lr = 0.008f;
        const float reg = 0.04f;
        const int epochs = 2;
        std::vector<float> pu(latent_dim);
        std::vector<float> qi(latent_dim);

        for (int epoch = 0; epoch < epochs; ++epoch) {
            for (const Rating& r : seen) {
                if (r.user < 0 || r.user >= users || r.item < 0 || r.item >= items) {
                    continue;
                }
                float* p_row = P + static_cast<long long>(r.user) * latent_dim;
                float* q_row = Q + static_cast<long long>(r.item) * latent_dim;
                float pred = global_mean;
                for (int k = 0; k < latent_dim; ++k) {
                    pu[k] = p_row[k];
                    qi[k] = q_row[k];
                    pred += pu[k] * qi[k];
                }
                const float err = r.rating - pred;
                for (int k = 0; k < latent_dim; ++k) {
                    p_row[k] += lr * (err * qi[k] - reg * pu[k]);
                    q_row[k] += lr * (err * pu[k] - reg * qi[k]);
                }
            }
        }
    }

    float predict(int user_id, int item_id) {
        if (user_id < 0 || user_id >= users || item_id < 0 || item_id >= items) {
            return global_mean;
        }
        float score = global_mean;
        const float* p_row = P + static_cast<long long>(user_id) * latent_dim;
        const float* q_row = Q + static_cast<long long>(item_id) * latent_dim;
        for (int k = 0; k < latent_dim; ++k) {
            score += p_row[k] * q_row[k];
        }
        return std::min(5.0f, std::max(0.5f, score));
    }

private:
    int users = 0;
    int items = 0;
    int latent_dim = 0;
    float global_mean = 0.0f;
    float* P = nullptr;
    float* Q = nullptr;
    std::vector<Rating> seen;
};

File: task2/track1_run/track1/students/solution.py
import numpy as np


class IncrementalSVD:
    """Memory-bounded incremental SGD reference solution."""

    def __init__(self, num_users, num_items, latent_dim, global_mean):
        self.num_users = num_users
        self.num_items = num_items
        self.latent_dim = latent_dim
        self.global_mean = float(global_mean)
        self.user_matrix = None
        self.item_matrix = None
        self.learning_rate = 0.0005
        self.regularization = 0.04
        self.local_epochs = 1
        self.chunk_size = 5_000

    def load_base_model(self, user_matrix: np.ndarray, item_matrix: np.ndarray):
        self.user_matrix = user_matrix
        self.item_matrix = item_matrix

    def update(self, incremental_batch: np.ndarray):
        if incremental_batch.size == 0:
            return

        ratings = np.asarray(incremental_batch, dtype=np.float32)
        lr_reg = self.learning_rate * self.regularization

        for _ in range(self.local_epochs):
            for start in range(0, ratings.shape[0], self.chunk_size):
                end = start + self.chunk_size
                chunk = ratings[start:end]
                u = chunk[:, 0].astype(np.int32)
                i = chunk[:, 1].astype(np.int32)
                r = chunk[:, 2]

                pu = self.user_matrix[u].copy()
                qi = self.item_matrix[i].copy()
                pred = self.global_mean + np.einsum("ij,ij->i", pu, qi)
                err = np.clip(r - pred, -5.0, 5.0).astype(np.float32)

                p_delta = self.learning_rate * (
                    err[:, None] * qi - self.regularization * pu
                )
                q_delta = self.learning_rate * err[:, None] * pu - lr_reg * qi
                np.add.at(self.user_matrix, u, p_delta)
                np.add.at(self.item_matrix, i, q_delta)

    def predict(self, user_id, item_id) -> float:
        if user_id < 0 or user_id >= self.num_users or item_id < 0 or item_id >= self.num_items:
            return self.global_mean
        score = self.global_mean + float(np.dot(self.user_matrix[user_id], self.item_matrix[item_id]))
        return min(5.0, max(0.5, score))
